File: java/src/main/java/problems101_200/LinkedListCycle.java
package problems101_200;

import concepts.SinglyListNode;
import java.util.HashSet;
import java.util.Set;

enum LinkedListCycle {
    REMEMBER_VISITS {
        @Override
        boolean hasCycle(SinglyListNode head) {
            // visited nodes
            Set<SinglyListNode> visits = new HashSet<>();
            for (SinglyListNode node : head) {
                if (!visits.add(node)) {
                    return true;
                }
            }
            return false;
        }
    },
    FAST_SLOW_POINTERS {
        @Override
        boolean hasCycle(SinglyListNode head) {
            if (head == null || head.next == null) {
                return false;
            }
            SinglyListNode slow = head.next;
            SinglyListNode fast = head.next.next;
            while (fast != null && fast.next != null && slow != fast) {
                slow = slow.next;
                fast = fast.next.next;
            }
            return slow == fast;
        }
    };

    abstract boolean hasCycle(SinglyListNode head);
}

File: kotlin/src/main/kotlin/problems101_200/LinkedListCycle.kt
package problems101_200

import concepts.SinglyListNode

enum class LinkedListCycle {
    REMEMBER_VISITS {
        override fun hasCycle(head: SinglyListNode?): Boolean {
            // visited nodes
            val visits = hashSetOf<SinglyListNode>()
            for (node in head!!) {
                if (!visits.add(node)) {
                    return true
                }
            }
            return false
        }
    },
    FAST_SLOW_POINTERS {
        override fun hasCycle(head: SinglyListNode?): Boolean {
            if (head?.next == null) {
                return false
            }
            var slow = head.next
            var fast = head.next!!.next
            while (fast != null && fast.hasNext() && slow !== fast) {
                slow = slow!!.next
                fast = fast.next!!.next
            }
            return slow === fast
        }
    },
    ;

    abstract fun hasCycle(head: SinglyListNode?): Boolean
}
